open-metadata-implementation/access-services/data-engine/data-engine-server/src/main/java/org/odpi/openmetadata/accessservices/dataengine/server/handlers/DataEngineConnectionAndEndpointHandler.java

```diff
@@ -0,0 +1,189 @@
+                CONNECTION_TYPE_NAME, qualifiedName, QUALIFIED_NAME_PROPERTY_NAME, connectionBuilder, methodName);
+    }
+
+    private void upsertEndpoint(String protocol, String networkAddress, String externalSourceGuid, String externalSourceName,
+                                String userId, String methodName, String connectionGuid, String endpointQualifiedName)
+            throws UserNotAuthorizedException, PropertyServerException, InvalidParameterException {
+        String endpointGuid;
+        Optional<EntityDetail> optionalEndpoint = dataEngineCommonHandler.findEntity(userId, endpointQualifiedName, ENDPOINT_TYPE_NAME);
+        if(optionalEndpoint.isPresent()){
+            updateEndpoint(optionalEndpoint.get(), protocol, networkAddress, externalSourceGuid, externalSourceName, userId, methodName);
+            endpointGuid = optionalEndpoint.get().getGUID();
+        } else {
+            endpointGuid = createEndpoint(protocol, networkAddress, endpointQualifiedName, externalSourceGuid,
+                    externalSourceName, userId, methodName);
+        }
+        dataEngineCommonHandler.upsertExternalRelationship(userId, endpointGuid, connectionGuid,
+                CONNECTION_ENDPOINT_TYPE_NAME, ENDPOINT_TYPE_NAME, externalSourceName, null);
+    }
+
+    private String createEndpoint(String protocol, String networkAddress, String qualifiedName, String externalSourceGuid,
+                                  String externalSourceName, String userId, String methodName)
+            throws InvalidParameterException, PropertyServerException, UserNotAuthorizedException {
+
+        EndpointBuilder endpointBuilder = new EndpointBuilder(protocol, networkAddress, qualifiedName, ENDPOINT_TYPE_GUID,
+                ENDPOINT_TYPE_NAME, repositoryHelper, serviceName, serverName);
+        return endpointHandler.createBeanInRepository(userId, externalSourceGuid, externalSourceName, ENDPOINT_TYPE_GUID,
+                ENDPOINT_TYPE_NAME, qualifiedName, QUALIFIED_NAME_PROPERTY_NAME, endpointBuilder, methodName);
+    }
+
+    private void updateEndpoint(EntityDetail endpoint,String protocol, String networkAddress, String externalSourceGuid,
+                                String externalSourceName, String userId, String methodName)
+            throws InvalidParameterException, PropertyServerException, UserNotAuthorizedException {
+        PrimitivePropertyValue protocolAsPropertyValue = new PrimitivePropertyValue();
+        protocolAsPropertyValue.setPrimitiveDefCategory(PrimitiveDefCategory.OM_PRIMITIVE_TYPE_STRING);
+        protocolAsPropertyValue.setPrimitiveValue(protocol);
+
+        PrimitivePropertyValue networkAddressAsPropertyValue = new PrimitivePropertyValue();
+        networkAddressAsPropertyValue.setPrimitiveDefCategory(PrimitiveDefCategory.OM_PRIMITIVE_TYPE_STRING);
+        networkAddressAsPropertyValue.setPrimitiveValue(networkAddress);
+
+        InstanceProperties properties = endpoint.getProperties();
```

Comment:
You can create an EndpointBuilder and get the properties using method getInstanceProperties instead of constructing the properties manually here